vllm/model_executor/models/mllama.py

```
@@ -169,6 +170,26 @@ def get_dummy_processor_inputs(
 class MllamaMultiModalProcessor(EncDecMultiModalProcessor[MllamaProcessingInfo]
                                 ):
 
+    def apply(
+        self,
+        prompt: Union[str, list[int]],
+        mm_data: MultiModalDataDict,
+        hf_processor_mm_kwargs: Mapping[str, object],
+    ) -> MultiModalEncDecInputs:
+        mm_inputs = super().apply(prompt, mm_data, hf_processor_mm_kwargs)
+
+        # Check that the number of image tokens matches the number of images
+        num_image_tokens = mm_inputs['prompt_token_ids'].count(
+            self.info.get_hf_config().image_token_index)
+        image_data = mm_data.get("image", [])
+        num_images = 1 if isinstance(image_data, Image) else len(image_data)
+        if num_image_tokens != num_images:
+            raise ValueError(
+                f"The number of image tokens ({num_image_tokens}) must be"
+                f" the same as the number of images ({num_images})")
```

Actually, we already check this via `BaseMultiModalProcessor._validate_mm_kwargs`, so I think this is no longer necessary...

---

The crash is reproducible on `main` using the request in https://github.com/vllm-project/vllm/pull/11939.

Looking at `_validate_mm_placeholders`, it seems to be a check for the placeholders generated for the encoder "prompt replacements", not a check on the number of image tokens going it to the decoder.

---

When you mention "crash", does it occur inside the processor or inside the model? If there is a mismatch, it should throw an error in `_validate_mm_placeholders` after calling `_find_mm_placeholders` (in case prompt replacement is done in HF) or `_apply_prompt_replacements` (in case prompt replacement is done in vLLM).

---

Thank you for the quick review and guidance!

> When you mention "crash", does it occur inside the processor or inside the model?

In the modeling code, it triggers [this assert statement](https://github.com/vllm-project/vllm/blob/d08b285adfb03f0b7952c8af4c2c83bb2dafcc2d/vllm/model_executor/models/mllama.py#L1510) when it tries to generate a cross-attention mask for an image token group that has no tiles / image embeddings.

In my (limited) understanding, all of the "prompt replacements" and "placeholders" are for injecting additional tokens corresponding to the image embeddings. In the `EncDecMultiModalProcessor`, the `BaseMultiModalProcessor` is only called with the encoder prompt ([REF](https://github.com/vllm-project/vllm/blob/d08b285adfb03f0b7952c8af4c2c83bb2dafcc2d/vllm/multimodal/processing.py#L1319-L1324)). `create_encoder_prompt` currently inspects the image data when constructing the encoder prompt ([REF](https://github.com/vllm-project/vllm/blob/d08b285adfb03f0b7952c8af4c2c83bb2dafcc2d/vllm/model_executor/models/mllama.py#L236-L244)). Should `create_encoder_prompt` be changed to count the number of image tokens in the prompt instead?